sycl/include/sycl/vector_preview.hpp

```diff
@@ -983,13 +983,21 @@ template <typename Type, int NumElements> class vec {
     }
   }
 
-#ifdef __SYCL_BINOP
-#error "Undefine __SYCL_BINOP macro"
+#if defined(__SYCL_BINOP) || defined(BINOP_BASE)
+#error "Undefine __SYCL_BINOP and BINOP_BASE macro"
 #endif
 
-#ifdef __SYCL_USE_EXT_VECTOR_TYPE__
-#define __SYCL_BINOP(BINOP, OPASSIGN, CONVERT)                                 \
-  friend vec operator BINOP(const vec &Lhs, const vec &Rhs) {                  \
+// Note: vec<>/SwizzleOp logical value is 0/-1 logic, as opposed to 0/1 logic.
+// As far as CTS validation is concerned, 0/-1 logic also applies when
+// NumElements is equal to one, which is somewhat inconsistent with being
+// transparent with scalar data.
+// TODO: Determine if vec<, NumElements=1> is needed at all, remove this
+// inconsistency if not by disallowing one-element vectors (as in OpenCL)
+#ifdef __SYCL_DEVICE_ONLY__
+#define BINOP_BASE(BINOP, OPASSIGN, CONVERT, COND)                             \
+  template <typename T = DataT>                                                \
```

Comment:
As an alternative to this approach, could we consider making a base-class hierarchy that we specialize for byte? I.e. something like:

```
vec<byte> -> detail::vec_arith<byte> ->detail::vec_arith_common
vec<int> -> detail::vec_arith<int> -> detail::vec_arith_common
```
Where `detail::vec_arith` would be specialized for types that have some operations that not all other vecs have and `detail::vec_arith_common` would have the operations _all_ `vec`'s have.

This way, we avoid making these operators dependent, which they aren't in the spec. Granted we do that in a number of places, but we might as well avoid it when we can.

Reply:
That can be done, but it looks a bit overkill, since the specialization is needed mostly for `std::byte`.  `enable_if_t<!detail::isbyte_v<T>>` would result in a more compact and readable code, then the base-class hierarchy. Nevertheless, if we decide to use the base-class hierarchy, we would also have to do that for `Swizzles`.

Tagging @aelovikov-intel for additional feedback on base-class hierarchy vs. `enable_if_t` approach.

Reply:
> This way, we avoid making these operators dependent, which they aren't in the spec.

I'm not sure what you meant by this? In the SPEC (https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_vec_interface), the math operations on `sycl::vec` are dependent/restricted to the under-lying data type.

Reply:
The class is dependent, the operators are not, which is why you have to add the templates here in order to use `std::enable_if_t`. In the spec, the operators are not defined as templated functions, so even if there is a default argument for each template argument, we technically deviate from the spec by adding them.

Reply:
> The class is dependent, the operators are not, which is why you have to add the templates here in order to use `std::enable_if_t`. In the spec, the operators are not defined as templated functions, so even if there is a default argument for each template argument, we technically deviate from the spec by adding them.

That makes sense. In that case, I think your base-class hierarchy approach would be better - I'll make that change.

Reply:
@steffenlarsen I've separated out all the math operations on `sycl::vec` in a base class hierarchy to specialize for `std::byte`. Note that now the operators are no longer friends of `sycl::vec` (which is required by the SPEC), but, I'm not sure how stringent that requirement is.

Reply:
It is a good point. I think it's up to interpretation. I don't believe the user should be able to experience this discrepancy, unlike having them templated.